### PR TITLE
Feature/plugin filter editor

### PR DESCRIPTION
**Pull request by infacc**

### CODE REVIEW

File: src/app/components-small/plugin-filter-editor/plugin-filter-editor.component.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        navigator.clipboard.writeText(this.filterString);
+    }
+
+    updateFilterEditor() {
+        const filterValue = this.filterControl.value;
+        if (filterValue == null) {
+            return;
+        }
+        if (this.filterControl.valid) {
+            this.filterObject = JSON.parse(filterValue);
```
Comment (buehlefs) [inline src/app/components-small/plugin-filter-editor/plugin-filter-editor.component.ts:51]: catch JSON paring errors that could happen here

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts [hunk 2]
```diff
@@ -36,47 +39,47 @@ export class PluginFilterNodeComponent implements OnInit {
             console.warn("No filter object provided to plugin filter node component");
             return;
         }
-        const isLeaf = (filter.name != null) || filter.tag || filter.version;
+        // Should filters have multiple attributes at some point, this must be changed
+        const type = Object.keys(filter)[0];
```
Comment (buehlefs) [inline src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts:43]: If you want to, you could add an active warning to the code:
```suggestion
        const filterKeys = Object.keys(filter)
        if (filterKeys.length != 1) {
            console.error("Filters with more than one attribute are not supported!");
        }
        const type = filterKeys[0];
```

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts [hunk 3]
```diff
@@ -36,47 +39,47 @@ export class PluginFilterNodeComponent implements OnInit {
...
+        if (type !== 'and' && type !== 'or' && type !== 'name' && type !== 'tag' && type !== 'version') {
```
Comment (buehlefs) [inline src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts:44]: Create a static set of allowed filter types and test if type is in that set (or refactor this test out into a method).
Even if this is probably the most optimized version it is also hard to read and maintain.

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts [hunk 4]
```diff
@@ -36,47 +39,47 @@ export class PluginFilterNodeComponent implements OnInit {
...
+            this.children = filter[this.type];
         }
         this.filterObject = currentValue;
         this.childChange.emit([this.index, this.filterObject]);
     }
 
-    setFilter(type: 'name' | 'tag' | 'version' | 'and' | 'or') {
-        const isLeaf = type === 'name' || type === 'tag' || type === 'version';
+    setFilter(type: FilterType) {
+        const isLeaf = type !== 'and' && type !== 'or';
```
Comment (buehlefs) [inline src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts:60]: It is probably a good idea to add a comment to remind future developers why this (and other locations) do not test for the "not" filter. The best location for this is probably the definition of the type "FilterType"

File: src/app/components-small/template-details/template-details.component.html [hunk 5]
```diff
@@ -20,7 +20,7 @@ <h2>Template Tab</h2>
         <mat-label>Sort Key:</mat-label>
         <input matInput type="number" formControlName="sortKey">
     </mat-form-field>
-    <qhana-plugin-filter-editor #filterEditor [tabLink]="tabLink"></qhana-plugin-filter-editor>
+    <qhana-plugin-filter-editor [tabLink]="tabLink" (filterEmitter)="onFilterChange($event)"></qhana-plugin-filter-editor>
```
Comment (buehlefs) [inline src/app/components-small/template-details/template-details.component.html:23]:
```suggestion
    <qhana-plugin-filter-editor [tabLink]="tabLink" (filterEmitter)="filterString = $event"></qhana-plugin-filter-editor>
```

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.html [hunk 6]
```diff
@@ -0,0 +1,71 @@
+<div>
+    <div *ngIf="isFilterEmpty()" class="create-buttons">
```
[the remark this answers is not in the document]
Comment (infacc) [reply src/app/components-small/plugin-filter-node/plugin-filter-node.component.html]: This is meant for empty filters (`{}`). I.e. when a tab is newly created. The HTML class name `create-buttons` is misleading, I changed it to `set-buttons`.

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts [hunk 7]
```diff
@@ -26,25 +27,23 @@ export class PluginFilterNodeComponent implements OnInit {
...
-        this.isEmpty = Object.keys(this.filterObject).length === 0;
-        if (this.isEmpty) {
-            return;
-        }
-        const currentValue = this.filterObject;
-        const filter = currentValue.not ?? this.filterObject;
-        this.inverted = currentValue.not != null;
+        this.isEmpty = Object.keys(this.filterIn).length === 0;
+        this.filterObject = { ...this.filterIn };
+        const filter = this.filterObject.not ?? this.filterObject;
```
[the remark this answers is not in the document]
Comment (infacc) [reply src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts]: Now deep copies are used. E.g. https://github.com/UST-QuAntiL/qhana-ui/blob/16dd6a53a3e6ab32cabb7e66e0f2e14998122ae0/src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts#L42

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts [hunk 8]
```diff
@@ -128,21 +129,21 @@ export class PluginFilterNodeComponent implements OnInit {
             this.value = null;
             filter[type] = this.children ?? [];
         }
-        this.childChange.emit(this.filterObject);
+        this.filterOut.emit(filter);
```
[the remark this answers is not in the document]
Comment (infacc) [reply src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts]: That wasn't intentional and is now fixed.

File: src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (buehlefs) [reply src/app/components-small/plugin-filter-node/plugin-filter-node.component.ts]: This should work. Shallow copies would also have worked if they cover every level of nesting that is inside the same filter node, as each node does these copies.

**Review by buehlefs — COMMENTED** [review]
Looks good, but there is another bug with the tab editing form. The form does not contain the name or description of an existing tab (the sort key may have the same problem)

**Comment by infacc** [comment]
The form didn't display the tab name/description/sortKey since I forgot to remove the filterString from the FormGroup. Now it works.

**Comment by infacc** [comment]
I added 2 more things:
- A reload button for the filter editor, which resets the filter editor to the saved filter.
- A JSON validator for the JSON editor (of the filter string). This doesn't validate the filter string but complains if the user enters invalid JSON.

File: src/app/components-small/plugin-filter-editor/plugin-filter-editor.component.html [hunk 9]
```diff
@@ -1,8 +1,14 @@
...
-    <mat-slide-toggle [(ngModel)]="showEditor">
-        Editor Mode {{ showEditor ? '(UI)' : '(JSON)' }}
-    </mat-slide-toggle>
+    <div class="editor-header">
+        <mat-slide-toggle [(ngModel)]="showEditor">
+            Editor Mode {{ showEditor ? '(UI)' : '(JSON)' }}
+        </mat-slide-toggle>
+        <button mat-raised-button type="button" (click)="ngOnInit()">
+            <mat-icon>refresh</mat-icon>
+            Reload Filter
```
Comment (buehlefs) [inline src/app/components-small/plugin-filter-editor/plugin-filter-editor.component.html:9]:
```suggestion
            Revert Changes
```

File: src/app/components-small/plugin-filter-editor/plugin-filter-editor.component.html [hunk 10]
```diff
@@ -1,8 +1,14 @@
...
     <h3 class="editor-description">Filter Editor</h3>
-    <mat-slide-toggle [(ngModel)]="showEditor">
-        Editor Mode {{ showEditor ? '(UI)' : '(JSON)' }}
-    </mat-slide-toggle>
+    <div class="editor-header">
+        <mat-slide-toggle [(ngModel)]="showEditor">
+            Editor Mode {{ showEditor ? '(UI)' : '(JSON)' }}
+        </mat-slide-toggle>
+        <button mat-raised-button type="button" (click)="ngOnInit()">
+            <mat-icon>refresh</mat-icon>
```
Comment (buehlefs) [inline src/app/components-small/plugin-filter-editor/plugin-filter-editor.component.html:8]: This is a decorative icon, so it should be hidden from screen readers: <https://material.angular.io/components/icon/overview#decorative-icons>

**Review by buehlefs — CHANGES_REQUESTED** [review]
Just one cosmetic/formulation change

**Review by buehlefs — APPROVED** [review]

**Pull request merged**